### PR TITLE
Implemented ExcludingMembersNamed

### LINKED ISSUE #182: [API Proposal]: ExcludingMemberNames(params string[] memberNames)
### Background and motivation

To globally exclude member names from an equivalency assertion of a complex graph, one can use the `IMemberInfo` overload as follows:

```csharp
var ignoredPropertyNames = new[] { "Parent", "Children" };
actual.Should().BeEquivalentTo(expected, config => config
    .Excluding((IMemberInfo member) => ignoredPropertyNames.Contains(member.Name)));
```

Please add a shortcut method that enables easily adding a list of properties to be excluded throughout the graph regardless of the type they belong to.

### API Proposal

```C#
public abstract class SelfReferenceEquivalencyOptions<TSelf> : IEquivalencyOptions
{
    public TSelf ExcludingMemberNames(params IEnumerable<string> excludedMemberNames) =>
        Excluding((IMemberInfo member) => excludedMemberNames.Contains(member.Name));
}
```

### API Usage

```C#
actual.Should().BeEquivalentTo(expected, options => options.ExcludingMemberNames("Parent", "Children"));
```

### Alternative Designs

_No response_

### Risks

_No response_

### Are you willing to help with a proof-of-concept (as PR in that or a separate repo) first and as pull-request later on?

Yes, please assign this issue to me.

### API Proposal

```C#
public class EnumAssertions<TEnum, TAssertions>
{
    public AndConstraint<TAssertions> BeDefined(string because = "", params object[] becauseArgs);
    public AndConstraint<TAssertions> NotBeDefined(string because = "", params object[] becauseArgs);
}
```

### API Usage

```C#
var dayOfWeek = (DayOfWeek)1;
dayOfWeek.Should().BeDefined();

var invalidDayOfWeek = (DayOfWeek)999;
invalidDayOfWeek.Should().NotBeDefined();
```

### Alternative Designs

_No response_

### Risks

_No response_

### Are you willing to help with a proof-of-concept (as PR in that or a separate repo) first and as pull-request later on?

No

### PR DESCRIPTION
**Pull request by weitzhandler**
Fixes #182 

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/main/Tests/AwesomeAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/main/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://awesomeassertions.org/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/main/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/main/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/main/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://awesomeassertions.org/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

**Comment by IT-VBFK** [comment]
Thank you for your contribution! 🎉

Out of curiosity: The same as proposed as in FA?

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        MemberSelectionContext context)
+    {
+        return selectedMembers.Where(p => !predicate(p.Expectation.Name)).ToArray();
+    }
+
+    /// <inheritdoc />
+    /// <filterpriority>2</filterpriority>
+    public override string ToString()
+    {
+        return "Exclude members named: " + description;
```
Comment (IT-VBFK) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs]:
```suggestion
        return $"Exclude members named: {description}";
```

This leverages the `null` check to the interpolated string handler which should increase the coverage

**Review by IT-VBFK — COMMENTED** [review]
Nice work

Just one small suggestion from my side

**Comment by IT-VBFK** [comment]
Ah yes.. just forgot: also please take care of the release notes, api snapshots and (if applicable) the docs for this change.

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs [hunk 2]
```diff
@@ -386,6 +386,22 @@ public TSelf ExcludingMissingMembers()
         return (TSelf)this;
     }
 
+    /// <summary>
+    /// Excludes the members with the specified names from the structural equality check.
+    /// </summary>
+    public TSelf ExcludingMemberNames(params string[] memberNames)
```
Comment (cbersch) [inline Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs]: I prefer `ExcludingMembersNamed`.

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs [hunk 3]
```diff
@@ -386,6 +386,22 @@ public TSelf ExcludingMissingMembers()
...
+    {
+        Guard.ThrowIfArgumentIsNull(memberNames);
+
+        if (memberNames.Length == 0)
```
Comment (cbersch) [inline Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs]: We shouldn't allow empty member names:
```C#
Guard.ThrowIfArgumentIsEmpty(memberNames);
```

**Review by cbersch — CHANGES_REQUESTED** [review]

**Comment by cbersch** [comment]
> Ah yes.. just forgot: also please take care of the release notes

Add 
```
## Unreleased

### Improvements
```
followed by the changes to `docs/_pages/releases.md`.

> the docs for this change.

I would add an example here: https://github.com/AwesomeAssertions/AwesomeAssertions/blob/aaa75a20a5cd645a6688531fcfaa8d36cd35b24e/docs/_pages/objectgraphs.md?plain=1#L168

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs (hunk 3 above)
Comment (weitzhandler) [reply Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs]: BTW, 

public static void ThrowIfArgumentIsEmpty<T>(IEnumerable<T> values, string paramName, string message)

Might be a good idea to use `CallerArgumentExpression` and add a default message.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using AwesomeAssertions.Equivalency;
+
+namespace AwesomeAssertions.Equivalency.Selection;
+
+/// <summary>
+/// Selection rule that removes a particular member from the structural comparison if its name matches the specified criteria.
+/// </summary>
+internal class ExcludeMemberByNameSelectionRule : IMemberSelectionRule
```
Comment (weitzhandler) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:11]: BTW, I think we could benefit from having an abstract `MemberSelectionRule` base implementation.

Comment (weitzhandler) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:11]: BTW @IT-VBFK and @cbersch, I think we could benefit from having an abstract `MemberSelectionRule` base implementation.

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs (hunk 3 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs]: The following check for `0` isn't necessary anymore.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 5]
```diff
@@ -166,6 +193,62 @@ public void When_excluding_members_it_should_fail_if_any_non_excluded_members_ar
...
+            };
+
+            // Act
+            Action act =
+                () => class1.Should().BeEquivalentTo(class2, opts => opts.ExcludingMembersNamed("Property1"));
+
+            // Assert
+            act.Should().Throw<XunitException>().WithMessage("Expected*Field3*");
+        }
+
```
Comment (cbersch) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: Please add one more test for no member names: `ExcludingMembersNamed()`

**Review by cbersch — CHANGES_REQUESTED** [review]

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 5 above)
Comment (ITaluone) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: And/or explicit `null` member names

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 4 above)
Comment (ITaluone) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:11]: Is there that much code to be shared?

You can open a new PR with your sketch

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+{
+    private readonly Func<string, bool> predicate;
+    private readonly string description;
+
+    public ExcludeMemberByNameSelectionRule(string[] memberNames)
```
Comment (IT-VBFK) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:16]: Should we extend this so that nested members can be excluded likewise?

Now you can only exclude top level members

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 5 above)
Comment (IT-VBFK) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: There are at least three tests missing:
- with empty `params` array
- with `null` `params` array
- with `params` which are `null` and `empty` mixed

Comment (cbersch) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: I wouldn't add tests for member names `new string[] { null }` and `new string [] { null, "" }`. Though this doesn't make sense, it seems overtesting to me.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 6 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:16]: I would expect the following example to succeed:
```C#
public class Foo
{
    public string Value { get; set; }
    public string DisplayName { get; set; }
    public Bar Nested { get; set; }
}

public class Bar
{
    public string Value { get; set; }
    public string DisplayName { get; set; }
}

Foo sut = new()
{
    Value = "1",
    DisplayName = "display1",
    Nested = new()
    {
        Value = "1.1",
        DisplayName = "display1.1"
    }
};
Foo expected = new()
{
    Value = "1",
    Nested = new()
    {
        Value = "1.1"
    }
};
sut.Should().BeEquivalentTo(expected, o => o.ExcludingMembersNamed("DisplayName"));
```
Can't test it at the moment.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 5 above)
Comment (IT-VBFK) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: Hmm

Yeah if you add them like this, it makes no sense, but if this array comes from some method dynamically we want to make sure nothing crashes

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 6 above)
Comment (weitzhandler) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:16]: @IT-VBFK Interesting. So we should iterate over all excluded names and treat them as individual expression trees? You mean that if we exclude "Nested.Value", a difference in "Value" should still fail, is that right? Any other scenarios?

@cbersch It does. But should it pass or fail the assertion if `DisplayName` is not excluded?

File: Src/AwesomeAssertions/Common/Guard.cs [hunk 7]
```diff
@@ -65,6 +65,16 @@ public static void ThrowIfArgumentContainsNull<T>(IEnumerable<T> values,
         }
     }
 
+    public static void ThrowIfArgumentContainsNullOrEmpty<T>(IEnumerable<IEnumerable<T>> values,
```
Comment (IT-VBFK) [inline Src/AwesomeAssertions/Common/Guard.cs:68]: I think we should just ignore `null` or empty values here. If we throw on every `null`/empty value, this can lead to tedious back-and-forth if you have more than one of such "empty" value

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 4 above)
Comment (weitzhandler) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:11]: We can always investigate and do at a later PR.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 6 above)
Comment (IT-VBFK) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:16]: If I understand @cbersch correctly, this should succeed.. so all good 👍

File: Src/AwesomeAssertions/Common/Guard.cs (hunk 7 above)
Comment (weitzhandler) [reply Src/AwesomeAssertions/Common/Guard.cs:68]: I'm not sure I follow desired behavior from above convesation, let's discuss there.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 5 above)
Comment (weitzhandler) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: ok guys let's decide what we disallow here a parameter, here are the options:

1. `memberNames` is `null`
2. `memberNames` is empty
3. `memberNames` contains a `null` or empty member name

Comment (IT-VBFK) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: >  with params which are null and empty mixed

for 3: I meant here, that we make sure, we do not pass `null` or empty down the excluding logic, because it makes no sense for me to do so.
I don't meant that it should throw (which is basically the point of such tests) if such scenarios appear.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 6 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:16]: > @IT-VBFK Interesting. So we should iterate over all excluded names and treat them as individual expression trees? You mean that if we exclude "Nested.Value", a difference in "Value" should still fail, is that right?

I would only check member names, not member paths, so excluding "Nested.Value" shouldn't have any impact.

> @cbersch It does. But should it pass or fail the assertion if `DisplayName` is not excluded?

Then it should fail.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        description = string.Join(", ", memberNames);
+        predicate = name => memberNames.Contains(name, StringComparer.Ordinal);
```
Comment (ITaluone) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:19]: IMHO there is some inconsistency between what the option's method is called `ExcludeMembersNamed` and what it really does: `.Contains`. I'm pretty sure that this shouldn't be the expected behavior:

```csharp
public class Foo
{
    public string Value { get; set; }
    public string DisplayName { get; set; }
    public Bar Nested { get; set; }
}

public class Bar
{
    public string Value { get; set; }
    public string DisplayName { get; set; }
}

Foo sut = new()
{
    Value = "1", 
    DisplayName = "display1", 
    Nested = new()
    {
        Value = "1.1", 
        DisplayName = "display1.1"
    }
};

Foo expected = new()
{
    Value = "1", 
    Nested = new()
    {
        Value = "1.1"
    }
};

sut.Should().BeEquivalentTo(expected, o => o.ExcludingMembersNamed("a", "e"));
```
This would in this case exclude _every_ member since every member has either an "a" or "e" (or both).

We need a little better logic, that simply call `.Contains()` I guess...

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 4 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:11]: Looking through all implementations I cannot see the benefits of an abstract base class.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 8 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:19]: It is `IEnumerable<string>.Contains`, which is fine here.
This examples fails, as expected with 
```
Expected property sut.DisplayName to be <null>, but found "display1".
Expected property sut.Nested.DisplayName to be <null>, but found "display1.1".
```

Comment (IT-VBFK) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:19]: Ou.. sorry :)

missed that fact

**Comment by IT-VBFK** [comment]
The discussion about `null` and empty elements is the only thing what's missing, correct?

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 9]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
...
+                Name = "Jack",
+                Age = 13
+            };
+
+            // Act
+            Action act = () => subject.Should().BeEquivalentTo(customer, options => options
+                .ExcludingMembersNamed("Name"));
+
+            // Assert
+            act.Should().NotThrow();
```
Comment (IT-VBFK) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:115]: We usually don't assert that something which shouldn't throw to `NotThrow`, because when it fails it fails anyways

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 10]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
...
+            };
+
+            // Act
+            Action act = () => subject.Should().BeEquivalentTo(subject, options => options
+                .ExcludingMembersNamed([]));
+
+            // Assert
+            act.Should().Throw<ArgumentException>()
+                .WithParameterName("memberNames")
+                .WithMessage("At least one member name must be specified.\nParameter name: memberNames");
```
Comment (IT-VBFK) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:183]: Just assert the shape of the message will do, like:

```suggestion
                .WithMessage("*one member*must be specified*memberNames*");
```

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 11]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
...
+            };
+
+            // Act
+            Action act = () => subject.Should().BeEquivalentTo(subject, options => options
+                .ExcludingMembersNamed([null, "Age"]));
+
+            // Assert
+            act.Should().Throw<ArgumentException>()
+                .WithParameterName("memberNames")
+                .WithMessage("Collection contains a null or empty value\nParameter name: memberNames");
```
Comment (IT-VBFK) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:202]: Same as above

File: docs/_pages/objectgraphs.md [hunk 12]
```diff
@@ -130,6 +130,12 @@ orderDto.Should().BeEquivalentTo(order, options =>
     options.Excluding(o => o.Products[1].Status));
 ```
 
+You can also exclude all members with of a specific name:
```
Comment (IT-VBFK) [inline docs/_pages/objectgraphs.md:133]:
```suggestion
You can also exclude all members with of a specific name anywhere in the object graph:
```

**Review by IT-VBFK — CHANGES_REQUESTED** [review]
Just a few notes which I had missed previously

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 9 above)
Comment (cbersch) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:115]: So this test would become
```C#
[Fact]
public void A_member_excluded_by_name_is_ignored()
{
    // Arrange
    var subject = new
    {
        Id = 1,
        Name = "John",
        Age = 13
    };

    var customer = new
    {
        Id = 1,
        Name = "Jack",
        Age = 13
    };

    // Act / Assert
    subject.Should().BeEquivalentTo(customer, options => options
        .ExcludingMembersNamed("Name"));
}
```
But this test isn't necessary at all, because it is a duplicate of `When_excluding_member_names_it_should_pass_if_only_the_excluded_members_are_different`.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 13]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
...
+                .ExcludingMembersNamed([]));
+
+            // Assert
+            act.Should().Throw<ArgumentException>()
+                .WithParameterName("memberNames")
+                .WithMessage("At least one member name must be specified.\nParameter name: memberNames");
+        }
+
+        [Fact]
+        public void When_an_empty_or_null_member_name_is_passed_in_memeber_names_parameter_it_should_throw()
```
Comment (cbersch) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:187]: typo: `_member_names_` (same in other method names)

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 14]
```diff
@@ -166,6 +306,62 @@ public void When_excluding_members_it_should_fail_if_any_non_excluded_members_ar
...
+                Property1 = "sit"
+            };
+
+            var class2 = new ClassWithSomeFieldsAndProperties
+            {
+                Field1 = "Lorem",
+                Field2 = "ipsum"
+            };
+
+            // Act
```
Comment (cbersch) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:328]:
```c#
// Act / Assert
class1.Should().BeEquivalentTo(class2,
    opts => opts.ExcludingMembersNamed("Field3", "Property1"));
```

File: docs/_pages/releases.md [hunk 15]
```diff
@@ -6,11 +6,15 @@ classes: wide
...
   nav: "sidebar"
 ---
+
 ## Unreleased
 
 ### What's new
 * Add new assertions for strings: `[Not]BeParsableInto` - [#185](https://github.com/AwesomeAssertions/AwesomeAssertions/pull/185)
 
+### Improvements
+* Add `ExcludingMembersNamed` to `BeEquivalentTo` which enables excluding members with specified name - [#183](https://github.com/AwesomeAssertions/AwesomeAssertions/pull/183)
```
Comment (cbersch) [inline docs/_pages/releases.md:16]: I would move that to `What's new`, sorry for the wrong proposal.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 5 above)
Comment (cbersch) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: > > with params which are null and empty mixed
> 
> for 3: I meant here, that we make sure, we do not pass `null` or empty down the excluding logic, because it makes no sense for me to do so. I don't meant that it should throw (which is basically the point of such tests) if such scenarios appear.

For me, it doesn't make sense to filter `null` or empty, as it doesn't help in any way. I also wouldn't throw in those cases. Only when `memberNames` are null or empty.

Comment (IT-VBFK) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:364]: Ok ok

Fine, if `IEnumberable<T>.Contains(null)` does nothing, because no member name can be `null`. So, basically "just" a unnecessary check, but won't hurt.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs (hunk 9 above)
Comment (IT-VBFK) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:115]: Correct

Comment (cbersch) [reply Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:115]: @IT-VBFK We actually have tests which use `Should().NotThrow()`. I guess those are older ones. Since we agree here, I'll change those tests to match our preferences.

**Comment by IT-VBFK** [comment]
@weitzhandler @cbersch Are there any undiscussed things left, or can we move on?

**Comment by cbersch** [comment]
> @weitzhandler @cbersch Are there any undiscussed things left, or can we move on?

For me only the unresolved conversations, with minor changes in tests.

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs [hunk 16]
```diff
@@ -386,6 +386,19 @@ public TSelf ExcludingMissingMembers()
         return (TSelf)this;
     }
 
+    /// <summary>
+    /// Excludes the members with the specified names from the structural equality check.
+    /// </summary>
+    public TSelf ExcludingMembersNamed(params string[] memberNames)
+    {
+        Guard.ThrowIfArgumentIsNull(memberNames);
+        Guard.ThrowIfArgumentIsEmpty(memberNames, nameof(memberNames), "At least one member name must be specified.");
```
Comment (lg2de) [inline Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs:395]: How about adding `exception` tag(s) for exceptions like this?
I thin it is useful to see (e.g. in IDE's tooltip) what is allowed and what is prohibited.

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public bool IncludesMembers => false;
+
+    public IEnumerable<IMember> SelectMembers(INode currentNode, IEnumerable<IMember> selectedMembers,
+        MemberSelectionContext context)
+    {
+        return selectedMembers.Where(p => !predicate(p.Expectation.Name)).ToArray();
```
Comment (lg2de) [inline Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:27]: Why do you think that creating an array is always good or maybe needed?
The method returns an `IEnumerable` only, this is what `Where` created itself.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 18]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
                 .WithMessage("*Exclude member when*Age*");
         }
 
+        [Fact]
+        public void A_member_excluded_by_name_is_described_in_the_failure_message()
```
Comment (lg2de) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:66]:
```suggestion
        public void Member_excluded_by_name_are_described_in_the_failure_message()
```

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 19]
```diff
@@ -62,6 +62,146 @@ public void A_member_excluded_by_predicate_is_described_in_the_failure_message()
...
+            // Act
+            Action act = () => subject.Should().BeEquivalentTo(customer, options => options
+                .ExcludingMembersNamed("Name"));
+
+            // Assert
+            act.Should().NotThrow();
+        }
+
+        [Fact]
+        public void A_member_excluded_by_name_should_pass_in_nested_types()
```
Comment (lg2de) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:119]: The test above is named "...should_not_fail" which is using negative statement.
Here you are using "should_pass" which is positive.
Generally I recommend using positive statements. At least you should be consistent.

File: Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs [hunk 20]
```diff
@@ -166,6 +306,62 @@ public void When_excluding_members_it_should_fail_if_any_non_excluded_members_ar
             act.Should().Throw<XunitException>().WithMessage("Expected*Field3*");
         }
 
+        [Fact]
+        [SuppressMessage("ReSharper", "StringLiteralTypo")]
```
Comment (lg2de) [inline Tests/AwesomeAssertions.Equivalency.Specs/SelectionRulesSpecs.Excluding.cs:310]: Instead of using "Lorem", "ipsum", etc. you could also use "A", "B", etc.

File: docs/_pages/releases.md (hunk 15 above)
Comment (lg2de) [inline docs/_pages/releases.md:16]:
```suggestion
* Add `ExcludingMembersNamed` to `BeEquivalentTo`'s options which enables excluding members with specified name - [#183](https://github.com/AwesomeAssertions/AwesomeAssertions/pull/183)
```

**Comment by IT-VBFK** [comment]
@weitzhandler What is the state of this one?

File: Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs (hunk 16 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/SelfReferenceEquivalencyOptions.cs:395]: Added this to the new PR

File: Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs (hunk 17 above)
Comment (cbersch) [reply Src/AwesomeAssertions/Equivalency/Selection/ExcludeMemberByNameSelectionRule.cs:27]: I think this is fine, otherwise the predicate is evaluated each time one iterates over the returned result.

**Comment by cbersch** [comment]
I should have addressed all open comments. Opened a new PR #317, this was stale for too long.
